core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java

```diff
@@ -45,49 +53,57 @@
     private final ShardId shard;
     private final boolean primary;
     private final String assignedNodeId;
-    private final Map<DiscoveryNode, Decision> nodeToDecision;
-    private final Map<DiscoveryNode, Float> nodeWeights;
     private final UnassignedInfo unassignedInfo;
     private final long remainingDelayNanos;
+    private final Set<String> activeAllocationIds;
+    private final Map<DiscoveryNode, NodeExplanation> nodeExplanations;
 
     public ClusterAllocationExplanation(ShardId shard, boolean primary, @Nullable String assignedNodeId,
-            UnassignedInfo unassignedInfo, Map<DiscoveryNode, Decision> nodeToDecision,
-            Map<DiscoveryNode, Float> nodeWeights, long remainingDelayNanos) {
+                                        UnassignedInfo unassignedInfo, Map<DiscoveryNode, Decision> nodeToDecision,
+                                        Map<DiscoveryNode, Float> nodeToWeight, long remainingDelayNanos,
+                                        List<IndicesShardStoresResponse.StoreStatus> shardStores, Set<String> activeAllocationIds) {
         this.shard = shard;
         this.primary = primary;
         this.assignedNodeId = assignedNodeId;
         this.unassignedInfo = unassignedInfo;
-        this.nodeToDecision = nodeToDecision == null ? Collections.emptyMap() : nodeToDecision;
-        this.nodeWeights = nodeWeights == null ? Collections.emptyMap() : nodeWeights;
         this.remainingDelayNanos = remainingDelayNanos;
+        this.activeAllocationIds = activeAllocationIds;
+        final Map<DiscoveryNode, Decision> nodeDecisions = nodeToDecision == null ? Collections.emptyMap() : nodeToDecision;
+        final Map<DiscoveryNode, Float> nodeWeights = nodeToWeight == null ? Collections.emptyMap() : nodeToWeight;
+        assert nodeDecisions.size() == nodeWeights.size() : "decision and weight list should be the same size";
+        final Map<DiscoveryNode, IndicesShardStoresResponse.StoreStatus> storeStatuses = calculateStoreStatuses(shardStores);
```

Comment:
I like that you separated the stats calculation from toXContent. I want to go one step further though (my previous comment was not explicit enough in that regard): Similar to other code (e.g. `NodeStats`), I want `ClusterAllocationExplanation` to just be a "result" object. This means that all the computation should be happening in `TransportClusterAllocationExplainAction`. The explanation does not need `activeAllocationIds`.